### PR TITLE
bpo-40939: Add the new grammar to the grammar specification documentation

### PR DESCRIPTION
**Pull request by pablogsal**
Example of how it looks like when using the new lexer:

grammar

https://bugs.python.org/issue40939

### CODE REVIEW

File: Doc/tools/extensions/peg_highlight.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+from pygments.lexer import RegexLexer, bygroups, include
+from pygments.token import (Comment, Generic, Keyword, Name, Operator,
+                            Punctuation, Text)
+
+from sphinx.highlighting import lexers
+
+class PEGLexer(RegexLexer):
+    """Pygments Lexer for PEG grammar (.gram) files
+
+    This lexer stripts the following elements from the grammar:
```
Comment (gvanrossum) [inline Doc/tools/extensions/peg_highlight.py]:
```suggestion
    This lexer strips the following elements from the grammar:
```

**Review by gvanrossum — COMMENTED** [review]
I worry about some of the rather verbose comments in the grammar file, and the various error rules. Also there are some cases where lookaheads are essential (notably the way formal parameters are now parsed). And maybe the ordering and formatting of some of the rules could be better.

PS. Your initial comments still shows variables (`a=...`) but the script correctly erases those.

**Comment by pablogsal** [comment]
> I worry about some of the rather verbose comments in the grammar file, and the various error rules.

I can try to eliminate alternatives if they start with "invalid_" but I think it will not be super resilient :(

> Also there are some cases where lookaheads are essential (notably the way formal parameters are now parsed).

Would you like to bring lookaheads back?

> And maybe the ordering and formatting of some of the rules could be better.

Not much we can do here with a custom lexer unless we touch the original file.

> PS. Your initial comments still shows variables (`a=...`) but the script correctly erases those.

Yeah, I removed the variables in the last amend after I realized that we wanted to get rid of them.

**Comment by gvanrossum** [comment]
> > I worry about some of the rather verbose comments in the grammar file, and the various error rules.
> 
> I can try to eliminate alternatives if they start with "invalid_" but I think it will not be super resilient :(

Plus, you'd have to eliminate both the alternatives and the rules. But these really do not deserve to be listed in the reference manual.

> > Also there are some cases where lookaheads are essential (notably the way formal parameters are now parsed).
> 
> Would you like to bring lookaheads back?

Only the ones that are significant -- and those in formal parameters *are* significant (without them, `def foo(bar baz)` would be legal)

> > And maybe the ordering and formatting of some of the rules could be better.
> 
> Not much we can do here with a custom lexer unless we touch the original file.

Yeah, we can do that later.

**Comment by pablogsal** [comment]
> Only the ones that are significant -- and those in formal parameters _are_ significant (without them, `def foo(bar baz)` would be legal)

Hmmm.... is there any obvious automatic way (using regular expressions) to detect the ones we want from the ones we don't want? Otherwise, we may need to maintain a copy of the grammar for displaying here and manually trim what we need.

**Comment by gvanrossum** [comment]
I dunno. Let's think about this some more before rushing into it. (We'd
have to explain a big about PEG's ordering of alternatives too, to guide
people using this as a basis for anything.)

**Comment by lysnikolaou** [comment]
> Hmmm.... is there any obvious automatic way (using regular expressions) to detect the ones we want from the ones we don't want?

A heuristic that is currently 100% correct is that a lookahead is only used for optimisation and can be skipped if and only if it is a positive lookahead and it appears first in its alternative. Although this is currently correct, it seems like something that could very easily be invalidated in the future, so I don't know if it can be used in the script.

**Review by willingc — COMMENTED** [review]
Overall, I like the approach. There are a few places where color makes a line choppy and hard to follow and may make it more difficult to read than just black text.

**Comment by gvanrossum** [comment]
I think I'm going to do something with this after all. Let's see if we can't make the highlighter also remove `invalid_*` rules and comments.

**Comment by gvanrossum** [comment]
Okay, @lysnikolaou we may need your review because now it's both Pablo's and my code.

**Comment by gvanrossum** [comment]
Should we backport this to 3.9? (Minus the Grammar/Grammar delete. :-)

**Comment by pablogsal** [comment]
> Should we backport this to 3.9? (Minus the Grammar/Grammar delete. :-)

I think it makes sense to do the backport. What do you think @lysnikolaou ?

**Comment by gvanrossum** [comment]
Okay then @pablogsal what do you think of my version?

**Review by willingc — APPROVED** [review]
LGTM. Thanks @pablogsal and @gvanrossum.

File: Doc/reference/grammar.rst [hunk 2]
```diff
@@ -1,7 +1,8 @@
 Full Grammar specification
 ==========================
 
-This is the full Python grammar, as it is read by the parser generator and used
-to parse Python source files:
+This is the full Python grammar, derived directly from the grammar
```
Comment (pablogsal) [inline Doc/reference/grammar.rst:4]: Question: Should we add a small note/sentence here explaining the difference between the actual grammar and the version included here?

**Review by pablogsal — COMMENTED** [review]
> Okay then @pablogsal what do you think of my version?

I think is great :) I left a minor question, but feel free to ignore

Note: I cannot approve my own PR, so go ahead and merge!

File: Doc/tools/extensions/peg_highlight.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    name = "PEG"
+    aliases = ["peg"]
+    filenames = ["*.gram"]
+    _name = r"([^\W\d]\w*)"
+    _text_ws = r"(\s*)"
+
+    tokens = {
+        "ws": [(r"\n", Text), (r"\s+", Text), (r"#.*$", Comment.Singleline),],
+        "lookaheads": [
```
Comment (pablogsal) [inline Doc/tools/extensions/peg_highlight.py:29]: I think we discussed this previously, but the only lookahead we should ignore are the ones that start a rule (the ones that we use for performance), the others I think we want to keep

Comment (lysnikolaou) [reply Doc/tools/extensions/peg_highlight.py:29]: Yup, I was just about to say this. I think I'd prefer having to answer the question "Why are these lookaheads here?" rather than "Why is this grammar not working properly?".

> but the only lookahead we should ignore are the ones that start a rule

And we could limit this even further to only the *positive* lookaheads that start a rule.

File: Doc/tools/extensions/peg_highlight.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            (r"(!\(.+\)\s?)", bygroups(None)),
+        ],
+        "metas": [
+            (r"(@\w+ '''(.|\n)+?''')", bygroups(None)),
+            (r"^(@.*)$", bygroups(None)),
+        ],
+        "actions": [(r"{(.|\n)+?}", bygroups(None)),],
+        "strings": [
+            (r"'\w+?'", Keyword),
+            (r'"\w+?"', Keyword),
```
Comment (lysnikolaou) [inline Doc/tools/extensions/peg_highlight.py:42]: I feel that we'd better differentiate between hard and soft keywords. What about this?

```suggestion
            (r'"\w+?"', Keyword.Pseudo),
```

It would then look like this:

Screenshot 2020-07-27 at 5 01 54 PM

Although this is surely a discussion for somewhere else, I also feel that we should revisit the idea of having double quotes for soft keywords. I keep forgetting which is which.

**Comment by lysnikolaou** [comment]
> > Should we backport this to 3.9? (Minus the Grammar/Grammar delete. :-)
> 
> I think it makes sense to do the backport. What do you think @lysnikolaou ?

I also think that backporting this to 3.9 makes sense.

File: Doc/tools/extensions/peg_highlight.py (hunk 4 above)
Comment (gvanrossum) [reply Doc/tools/extensions/peg_highlight.py:42]: Let's put this discussion off until we actually *have* soft keywords (i.e. if/when PEP 622 is implemented).

**Comment by gvanrossum** [comment]
The new version I pushed has a longer intro in grammar.rst, both to explain (some of) what's omitted and to clarify the notation. It also restores most lookaheads, except for the positive ones that are purely optimizations (as a heuristic I use `|` followed by a positive lookahead). It makes the grammar a little uglier but reduces the possibility of looking like an ambiguity.

File: Doc/tools/extensions/peg_highlight.py (hunk 4 above)
Comment (lysnikolaou) [reply Doc/tools/extensions/peg_highlight.py:42]: Yup, agreed.

What about the difference in highlighting? Do you want to go with this now or should we also care about this once we have soft keywords?

Comment (gvanrossum) [reply Doc/tools/extensions/peg_highlight.py:42]: I don't think we should use a different style for soft keywords.

File: Doc/tools/extensions/peg_highlight.py (hunk 3 above)
Comment (gvanrossum) [reply Doc/tools/extensions/peg_highlight.py:29]: Done.

File: Doc/tools/extensions/peg_highlight.py (hunk 4 above)
Comment (lysnikolaou) [reply Doc/tools/extensions/peg_highlight.py:42]: Ok.

**Review by lysnikolaou — APPROVED** [review]
LGTM. Thanks!!
